### ISSUE TITLE
Field filter linked to Boolean column causes an error when refreshing a public question

### ISSUE THREAD

**Issue by noahmoss** [metabase/metabase#22591]
**Describe the bug**
If you create a field filter in a SQL question that is linked to a boolean column, you can use it in a public question when the page first loads, but if you subsequently refresh the page, it results in an error.

**Logs**

<details>

```
2022-05-10 14:42:19,795 ERROR middleware.catch-exceptions :: Error processing query: null
{:database_id 13,
 :started_at #t "2022-05-10T14:42:19.674907Z[UTC]",
 :via
 [{:status :failed,
   :class clojure.lang.ExceptionInfo,
   :error
   "Error executing query: ERROR: operator does not exist: boolean = character varying\n  Hint: No operator matches the given name and argument types. You might need to add explicit type casts.\n  Position: 189",
   :stacktrace
   ["--> driver.sql_jdbc.execute$execute_reducible_query$fn__68478.invoke(execute.clj:502)"
    "driver.sql_jdbc.execute$execute_reducible_query.invokeStatic(execute.clj:499)"
    "driver.sql_jdbc.execute$execute_reducible_query.invoke(execute.clj:486)"
    "driver.sql_jdbc.execute$execute_reducible_query.invokeStatic(execute.clj:494)"
    "driver.sql_jdbc.execute$execute_reducible_query.invoke(execute.clj:486)"
    "driver.sql_jdbc$eval79310$fn__79311.invoke(sql_jdbc.clj:54)"
    "query_processor.context$executef.invokeStatic(context.clj:59)"
    "query_processor.context$executef.invoke(context.clj:48)"
    "query_processor.context.default$default_runf.invokeStatic(default.clj:67)"
    "query_processor.context.default$default_runf.invoke(default.clj:65)"
    "query_processor.context$runf.invokeStatic(context.clj:45)"
    "query_processor.context$runf.invoke(context.clj:39)"
    "query_processor.reducible$identity_qp.invokeStatic(reducible.clj:12)"
    "query_processor.reducible$identity_qp.invoke(reducible.clj:9)"
    "query_processor.middleware.cache$maybe_return_cached_results$maybe_return_cached_results_STAR___59826.invoke(cache.clj:220)"
    "query_processor.middleware.permissions$check_query_permissions$fn__62733.invoke(permissions.clj:109)"
    "metabase_enterprise.advanced_permissions.query_processor.middleware.permissions$check_download_permissions$fn__66490.invoke(permissions.clj:124)"
    "metabase_enterprise.sandbox.query_processor.middleware.column_level_perms_check$maybe_apply_column_level_perms_check$fn__66734.invoke(column_level_perms_check.clj:30)"
    "query_processor.middleware.mbql_to_native$mbql__GT_native$fn__64027.invoke(mbql_to_native.clj:23)"
    "query_processor$fn__67270$combined_post_process__67275$combined_post_process_STAR___67276.invoke(query_processor.clj:207)"
    "query_processor$fn__67270$combined_pre_process__67271$combined_pre_process_STAR___67272.invoke(query_processor.clj:204)"
    "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__65244$fn__65249.invoke(resolve_database_and_driver.clj:35)"
    "driver$do_with_driver.invokeStatic(src/metabase/driver.clj:75)"
    "driver$do_with_driver.invoke(src/metabase/driver.clj:71)"
    "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__65244.invoke(resolve_database_and_driver.clj:34)"
    "query_processor.middleware.fetch_source_query$resolve_card_id_source_tables$fn__63834.invoke(fetch_source_query.clj:322)"
    "query_processor.middleware.store$initialize_store$fn__66078$fn__66079.invoke(store.clj:11)"
    "query_processor.store$do_with_store.invokeStatic(store.clj:44)"
    "query_processor.store$do_with_store.invoke(store.clj:38)"
    "query_processor.middleware.store$initialize_store$fn__66078.invoke(store.clj:10)"
    "query_processor.middleware.normalize_query$normalize$fn__64046.invoke(normalize_query.clj:22)"
    "metabase_enterprise.audit_app.query_processor.middleware.handle_audit_queries$handle_internal_queries$fn__66701.invoke(handle_audit_queries.clj:130)"
    "query_processor.middleware.constraints$add_default_userland_constraints$fn__53397.invoke(constraints.clj:52)"
    "query_processor.middleware.process_userland_query$process_userland_query$fn__65117.invoke(process_userland_query.clj:145)"
    "query_processor.middleware.catch_exceptions$catch_exceptions$fn__62826.invoke(catch_exceptions.clj:162)"
    "query_processor.reducible$async_qp$qp_STAR___55008$thunk__55010.invoke(reducible.clj:100)"
    "query_processor.reducible$async_qp$qp_STAR___55008$fn__55012.invoke(reducible.clj:105)"],
   :error_type :invalid-query,
   :ex-data
   {:sql
    "-- Metabase:: userID: 1 queryType: native queryHash: dc2feab6b427049aab0faeabe4906edda1490aaa63b07f1893009f5e57500354\nselect * from bool_table where \"public\".\"bool_table\".\"is_a_happy_row\" = ?",
    :params ["false"],
    :type :invalid-query}}],
 :state "42883",
 :error_type :invalid-query,
 :json_query
 {:constraints {:max-results 10000, :max-results-bare-rows 2000},
  :type :native,
  :middleware {:js-int-to-string? true, :ignore-cached-results? nil},
  :native
  {:template-tags
   {"is_happy"
    {:id "a1b0dfd2-fdfd-527b-e14a-125c398c26ea",
     :name "is_happy",
     :display-name "Is happy",
     :type :dimension,
     :dimension [:field 109 nil],
     :widget-type :string/=,
     :default [true]}},
   :query "select * from bool_table where {{is_happy}}"},
  :database 13,
  :parameters
  ({:type "string/=",
    :value ["false"],
    :target ["dimension" ["template-tag" "is_happy"]],
    :id "a1b0dfd2-fdfd-527b-e14a-125c398c26ea"}),
  :async? true,
  :cache-ttl nil},
 :status :failed,
 :class org.postgresql.util.PSQLException,
 :stacktrace
 ["org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2675)"
  "org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2365)"
  "org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:355)"
  "org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:490)"
  "org.postgresql.jdbc.PgStatement.execute(PgStatement.java:408)"
  "org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:166)"
  "org.postgresql.jdbc.PgPreparedStatement.executeQuery(PgPreparedStatement.java:118)"
  "com.mchange.v2.c3p0.impl.NewProxyPreparedStatement.executeQuery(NewProxyPreparedStatement.java:1471)"
  "--> driver.sql_jdbc.execute$eval68378$fn__68379.invoke(execute.clj:363)"
  "driver.sql_jdbc.execute$execute_statement_or_prepared_statement_BANG_.invokeStatic(execute.clj:376)"
  "driver.sql_jdbc.execute$execute_statement_or_prepared_statement_BANG_.invoke(execute.clj:372)"
  "driver.sql_jdbc.execute$execute_reducible_query$fn__68478.invoke(execute.clj:500)"
  "driver.sql_jdbc.execute$execute_reducible_query.invokeStatic(execute.clj:499)"
  "driver.sql_jdbc.execute$execute_reducible_query.invoke(execute.clj:486)"
  "driver.sql_jdbc.execute$execute_reducible_query.invokeStatic(execute.clj:494)"
  "driver.sql_jdbc.execute$execute_reducible_query.invoke(execute.clj:486)"
  "driver.sql_jdbc$eval79310$fn__79311.invoke(sql_jdbc.clj:54)"
  "query_processor.context$executef.invokeStatic(context.clj:59)"
  "query_processor.context$executef.invoke(context.clj:48)"
  "query_processor.context.default$default_runf.invokeStatic(default.clj:67)"
  "query_processor.context.default$default_runf.invoke(default.clj:65)"
  "query_processor.context$runf.invokeStatic(context.clj:45)"
  "query_processor.context$runf.invoke(context.clj:39)"
  "query_processor.reducible$identity_qp.invokeStatic(reducible.clj:12)"
  "query_processor.reducible$identity_qp.invoke(reducible.clj:9)"
  "query_processor.middleware.cache$maybe_return_cached_results$maybe_return_cached_results_STAR___59826.invoke(cache.clj:220)"
  "query_processor.middleware.permissions$check_query_permissions$fn__62733.invoke(permissions.clj:109)"
  "metabase_enterprise.advanced_permissions.query_processor.middleware.permissions$check_download_permissions$fn__66490.invoke(permissions.clj:124)"
  "metabase_enterprise.sandbox.query_processor.middleware.column_level_perms_check$maybe_apply_column_level_perms_check$fn__66734.invoke(column_level_perms_check.clj:30)"
  "query_processor.middleware.mbql_to_native$mbql__GT_native$fn__64027.invoke(mbql_to_native.clj:23)"
  "query_processor$fn__67270$combined_post_process__67275$combined_post_process_STAR___67276.invoke(query_processor.clj:207)"
  "query_processor$fn__67270$combined_pre_process__67271$combined_pre_process_STAR___67272.invoke(query_processor.clj:204)"
  "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__65244$fn__65249.invoke(resolve_database_and_driver.clj:35)"
  "driver$do_with_driver.invokeStatic(src/metabase/driver.clj:75)"
  "driver$do_with_driver.invoke(src/metabase/driver.clj:71)"
  "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__65244.invoke(resolve_database_and_driver.clj:34)"
  "query_processor.middleware.fetch_source_query$resolve_card_id_source_tables$fn__63834.invoke(fetch_source_query.clj:322)"
  "query_processor.middleware.store$initialize_store$fn__66078$fn__66079.invoke(store.clj:11)"
  "query_processor.store$do_with_store.invokeStatic(store.clj:44)"
  "query_processor.store$do_with_store.invoke(store.clj:38)"
  "query_processor.middleware.store$initialize_store$fn__66078.invoke(store.clj:10)"
  "query_processor.middleware.normalize_query$normalize$fn__64046.invoke(normalize_query.clj:22)"
  "metabase_enterprise.audit_app.query_processor.middleware.handle_audit_queries$handle_internal_queries$fn__66701.invoke(handle_audit_queries.clj:130)"
  "query_processor.middleware.constraints$add_default_userland_constraints$fn__53397.invoke(constraints.clj:52)"
  "query_processor.middleware.process_userland_query$process_userland_query$fn__65117.invoke(process_userland_query.clj:145)"
  "query_processor.middleware.catch_exceptions$catch_exceptions$fn__62826.invoke(catch_exceptions.clj:162)"
  "query_processor.reducible$async_qp$qp_STAR___55008$thunk__55010.invoke(reducible.clj:100)"
  "query_processor.reducible$async_qp$qp_STAR___55008$fn__55012.invoke(reducible.clj:105)"],
 :card_id 12,
 :context :public-question,
 :error
 "ERROR: operator does not exist: boolean = character varying\n  Hint: No operator matches the given name and argument types. You might need to add explicit type casts.\n  Position: 189",
 :row_count 0,
 :running_time 0,
 :data {:rows [], :cols []}}

```

</details>

**To Reproduce**
1. Create and save a SQL question with a field filter linked to a Boolean column
2. Create a public link for the question, and visit the link
3. The query should run successfully when initially visiting the link (with no query params in the URL)
4. If you refresh the page (with query params now appended to the URL) the query errors 

**Screenshots**

https://user-images.githubusercontent.com/32746338/167657093-69209d19-c49d-4cd8-944a-a5eca2a1bec6.mov

**Information about your Metabase Installation:**
Current master

**Additional context**
Seems to be an issue with inferring the type of the filter value when passed in via query params. Strongly related to #22552.

**Comment by flamber** [comment]
Looks like the same as #15498, which was fixed in 0.41.0, but now only a problem for Public question.
And seems it's a regression since 0.42 (cannot reproduce on 0.41.6, but can reproduce on 0.42.4).

**Comment by noahmoss** [comment]
Appears to have been introduced by #18431
